### PR TITLE
Reset Token position on reuse in `predicate_token_filter`

### LINKED ISSUE #47197: predicate_token_filter : The token.getPosition() method return wrong value
**Elasticsearch version** :
7.3.2 dockerized
image: docker.elastic.co/elasticsearch/elasticsearch:7.3.2

**Plugins installed**: none

**JVM version** (`java -version`):

openjdk version "12.0.2" 2019-07-16
OpenJDK Runtime Environment (build 12.0.2+10)
OpenJDK 64-Bit Server VM (build 12.0.2+10, mixed mode, sharing)

**OS version** (`uname -a` if on a Unix-like system):

Linux 7f94601adc38 4.20.7-042007-generic #201902061234 SMP Wed Feb 6 17:36:40 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
I'm using a predicate_token_filter to keep only the first X token of a stream. For this I use this filter configuration : 

`{
	"myPredicatefilter": {
		"type": "predicate_token_filter",
		"script": {
			"source": "token.getPosition() <= 1"
		}
	}
}`

But every time I use this analyzer it seems the position of the tokens are increasing, and after a few calls, the filter does not produce any token. 

[Here is a video showing the problem : ](https://streamable.com/1ag0f)

**Steps to reproduce**:

Complete index settings : 

```
PUT issue-predicate-token-filter
{
  "settings": {
    "analysis": {
      "filter": {
        "myPredicatefilter": {
          "type": "predicate_token_filter",
          "script": {
            "source": "token.getPosition() <= 1"
          }
        }
      },
      "analyzer": {
        "myPredicateAnalyzer": {
          "filter": [
            "myPredicatefilter"
          ],
          "type": "custom",
          "tokenizer": "whitespace"
        }
      }
    }
  }
}
```

**analyze request :** 

```
POST issue-predicate-token-filter/_analyze
{
  "analyzer": "myPredicateAnalyzer",
  "text": "pain grillé"
}
```
**first result :** 

```
{
  "tokens" : [
    {
      "token" : "pain",
      "start_offset" : 0,
      "end_offset" : 4,
      "type" : "word",
      "position" : 0
    },
    {
      "token" : "grillé",
      "start_offset" : 5,
      "end_offset" : 11,
      "type" : "word",
      "position" : 1
    }
  ]
}


```

**second result and all call afterward** : 

```
{
  "tokens" : [ ]
}
```

The analyzer can work again for a call after a **_close / _open** in the index. And also if I use `explain : true` in the analyze request the analyzer works without any problem.

```
POST issue-predicate-token-filter/_analyze
{
  "analyzer": "myPredicateAnalyzer",
  "text": "pain grillé",
  "explain": true
}
```
You can see the weird behavior by adding a Debug.explain in the filter script

```
PUT issue-predicate-token-filter-with-debug
{
  "settings": {
    "analysis": {
      "filter": {
        "myPredicatefilter": {
          "type": "predicate_token_filter",
          "script": {
            "source": "Debug.explain(token.getPosition())"
          }
        }
      },
      "analyzer": {
        "myPredicateAnalyzer": {
          "filter": [
            "myPredicatefilter"
          ],
          "type": "custom",
          "tokenizer": "whitespace"
        }
      }
    }
  }
}
```

```
POST issue-predicate-token-filter-with-debug/_analyze
{
  "analyzer": "myPredicateAnalyzer",
  "text": "pain grillé",
  "explain": false
}
```
You will see the token.getPosition() value increasing after each call.

### PR DESCRIPTION
**Pull request by romseygeek**
Most of the information in `AnalysisPredicateScript.Token` is pulled directly
from its underlying AttributeSource, but we also keep track of the token position,
and this state is held directly on the Token.  This information needs to be reset when
the containing `ScriptFilteringTokenFilter` is re-used.

Fixes #47197

### CODE REVIEW

**Comment by romseygeek** [comment]
@elasticmachine update branch

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/PredicateTokenFilterScriptFactory.java [hunk 1]
```diff
@@ -69,5 +69,11 @@ protected boolean accept() throws IOException {
             token.updatePosition();
             return script.execute(token);
         }
+
+        @Override
+        public void reset() throws IOException {
```
Comment (jimczi) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/PredicateTokenFilterScriptFactory.java]: I think `ScriptedConditionTokenFilterFactory` has the same issue ?

**Review by jimczi — COMMENTED** [review]
I left some comments but the fix looks good to me

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/PredicateTokenScriptFilterTests.java [hunk 2]
```diff
@@ -79,8 +79,8 @@ public boolean execute(Token token) {
 
         try (NamedAnalyzer analyzer = analyzers.get("myAnalyzer")) {
             assertNotNull(analyzer);
-            assertAnalyzesTo(analyzer, "Vorsprung Durch Technik", new String[]{
-                "Vorsprung", "Technik"
+            assertAnalyzesTo(analyzer, "Oh what a wonderful thing to be", new String[]{
```
[the remark this answers is not in the document]
Comment (romseygeek) [reply modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/PredicateTokenScriptFilterTests.java]: `assertAnalyzesTo` already checks reset behaviour - this test fails if the reset logic isn't present

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/PredicateTokenFilterScriptFactory.java (hunk 1 above)
Comment (romseygeek) [reply modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/PredicateTokenFilterScriptFactory.java]: ++, will update and add a test

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/PredicateTokenScriptFilterTests.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimczi) [reply modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/PredicateTokenScriptFilterTests.java]: got it thanks

**Pull request merged**